app/src/main/java/com/example/d3companion/presenters/list/ListPresenter.kt

```diff
@@ -1,21 +1,27 @@
 package com.example.d3companion.presenters.list
 
+import android.util.Log
 import com.example.d3companion.models.D3ViewElement
 import com.example.d3companion.models.D3ViewType
+import com.example.d3companion.services.D3IcyVeinsListener
+import com.example.d3companion.services.FileD3IcyVeinsProvider
 import com.example.d3companion.views.list.ListView
 import java.lang.ref.WeakReference
 import java.util.ArrayList
 
-class ListPresenter(view: ListView) : ListPresenterProvider {
+class ListPresenter(view: ListView, provider: FileD3IcyVeinsProvider) : ListPresenterProvider, D3IcyVeinsListener {
 
     private var weakView: WeakReference<ListView>? = null
+    private val dataProvider: FileD3IcyVeinsProvider = provider
```

Comment:
`dataProvider` needs to be of type `WeakReference<D3IcyVeinsProvider>?`, otherwise both provider and listener will have a strong reference to each other and, therefore, will not be collected by the garbage collector.